### PR TITLE
refactor(build script): add generate-dts script

### PR DESCRIPTION
**Pull request by agankarin**

### CODE REVIEW

**Comment by agankarin** [comment]
I'll update `docworks-dts` version the moment I'll merge this PR: https://github.com/wix/docworks/pull/18
Then the tests should pass

File: scripts/selector-declaration-builder.js [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+      `${WIX_SELECTOR_TYPE_VARIABLE}`,
+      `keyof ${WIX_SELECTOR_TYPE}`
+    )
+  ];
+  const jsDocComment = operation.docs.summary;
+
+  return { name, params, returnType, typeParameters, jsDocComment };
+}
+
+function convent$wOperationToFunctionType(operation) {
```
Comment (tomenden) [inline scripts/selector-declaration-builder.js]: Typo - should be `convert`

File: types/common/utilityTypes.d.ts [hunk 2]
```diff
@@ -0,0 +1,3 @@
+type IntersectionArrayAndBase<T, P> = {
```
[the remark this answers is not in the document]
Comment (tomenden) [reply types/common/utilityTypes.d.ts]: Can we move its definition to inside the $w-declaration? Are there advantages to leaving it separate and referencing it?

File: scripts/selector-declaration-generator.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    repo.services,
+    tripleSlashDirectivePath
+  );
+
+  const $wDtsFilePath = "types/common/$w.d.ts";
+  fs.writeFileSync($wDtsFilePath, $wDeclarationContent);
+  console.log(`selector dts saved to file ${$wDtsFilePath}`);
+}
+
+generate$wdtsFile();
```
Comment (tomenden) [inline scripts/selector-declaration-generator.js]:
```suggestion
generate$wDtsFile()
```

File: package.json [hunk 4]
```diff
@@ -4,11 +4,13 @@
...
   "scripts": {
     "lint": "eslint .",
-    "build": ":",
+    "build": "npm run generate-dts",
     "test": "jest",
-    "generate-dts": "docworks dts -r git@github.com:wix-private/wix-code-docs.git -o declaration -d temp_declarations --wixselector && npm run copy-declarations temp_declarations",
-    "postpublish" : "PACKAGE_VERSION=$(cat package.json | grep \\\"version\\\" | head -1 | awk -F: '{ print $2 }' | sed 's/[\",]//g' | tr -d '[[:space:]]') && git tag v$PACKAGE_VERSION && git push --tags",
-    "copy-declarations": "scripts/copy_declarations.sh"
+    "clone-repo": "rm -rf corvid_docs && git clone git@github.com:wix-private/wix-code-docs.git corvid_docs",
+    "generate-dts": "npm run clone-repo && npm run generate-dts:declaration && npm run generate-dts:selector && rm -rf corvid_docs",
```
Comment (tomenden) [inline package.json]: Consider making `corvid_docs` a parameter to all other scripts, so you only have to write it in this one line. For example:
```suggestion
    "generate-dts": "npm run clone-repo -- corvid_docs && npm run generate-dts:declaration -- corvid_docs && npm run generate-dts:selector -- corvid_docs && rm -rf corvid_docs",
```

This will require making some minor changes in the following places:
https://github.com/wix-incubator/corvid-types/blob/dae32d20ae425932558e99ebc15bf64c6bf766e9/package.json#L9
https://github.com/wix-incubator/corvid-types/blob/dae32d20ae425932558e99ebc15bf64c6bf766e9/package.json#L11
https://github.com/wix-incubator/corvid-types/blob/dae32d20ae425932558e99ebc15bf64c6bf766e9/scripts/selector-declaration-generator.js#L6


Alternatively (and perhaps a nice approach) is to create one `generate-dts.js` script that will orchestrate it all 🎼 🎻

File: scripts/selector-declaration-generator.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
+#!/usr/bin/env node
+const fs = require("fs");
+const { readFromDir } = require("docworks-repo");
+const $wDeclarationBuilder = require("./selector-declaration-builder");
+
+async function generate$wdtsFile() {
+  const corvidDocsPath = "corvid_docs";
```
Comment (tomenden) [inline scripts/selector-declaration-generator.js]: consider making `corvid_docs` a paremeter

File: test/it/negative.spec.js [hunk 6]
```diff
@@ -1,9 +1,16 @@
...
-const negativeRoots = listSubDirectories(NEGATIVE_ROOT_PATH);
+const ignoredTests = [
+  "test/it/code-samples/negative/pages-$w-dynamic-not-on-page"
+];
+const negativeRoots = listSubDirectories(NEGATIVE_ROOT_PATH).filter(
+  subDirectory => !ignoredTests.includes(subDirectory)
+);
 
 describe("typescript - negative scenarios", () => {
+  console.log(negativeRoots);
```
Comment (tomenden) [inline test/it/negative.spec.js]: Left here on purpose?

File: test/it/negative.spec.js [hunk 7]
```diff
@@ -1,9 +1,16 @@
 const compiler = require("../utils/compiler");
 const listSubDirectories = require("../utils/listSubDirectories");
+
 const NEGATIVE_ROOT_PATH = "test/it/code-samples/negative";
-const negativeRoots = listSubDirectories(NEGATIVE_ROOT_PATH);
+const ignoredTests = [
+  "test/it/code-samples/negative/pages-$w-dynamic-not-on-page"
```
Comment (tomenden) [inline test/it/negative.spec.js]: Do we have a jira for fixing this test? Can you add a link pls?

File: test/scripts/selector-declaration-builder.test.js [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    return $wDeclarationBuilder(services, tripleSlashDirectivePath);
+  }
+
+  describe("triple slash directive", () => {
+    test("should generate triple slash directive with the given path", () => {
+      const tripleSlashDirectivePath = "../common/utilityTypes.d.ts";
+      const dts = get$wDts({ tripleSlashDirectivePath });
+      const expectedDeceleration = ``; /// <reference path="${tripleSlashDirectivePath}" />`
```
Comment (tomenden) [inline test/scripts/selector-declaration-builder.test.js]: This is currently an empty string, I am assuming not on purpose

File: test/it/negative.spec.js (hunk 7 above)
Comment (agankarin) [reply test/it/negative.spec.js]: Thanks! Just add to our current sprint

File: .gitignore [hunk 9]
```diff
@@ -1,4 +1,6 @@
 node_modules
+corvid_docs
```
Comment (dor-itzhaki) [inline .gitignore]: what is corvid_docs ?

File: configs/tsconfig.backend.json [hunk 10]
```diff
@@ -2,6 +2,6 @@
   "extends": "./tsconfig.base.json",
   "compilerOptions": {
     "typeRoots": ["../", "node_modules/@types"],
-    "types": ["types/backend", "node"]
+    "types": ["types/backend"]
```
Comment (dor-itzhaki) [inline configs/tsconfig.backend.json]: why did you remove node types ?

File: src_types/common/utilityTypes.d.ts [hunk 11]
```diff
@@ -0,0 +1,3 @@
+type IntersectionArrayAndBase<T, P> = {
```
Comment (dor-itzhaki) [inline src_types/common/utilityTypes.d.ts]: last time i'll say it... :)
no reason to have this file. $w is the only one using it and probably the only one that'll ever use it (or something else eventually). no reason for this extra file.

File: test/it/negative.spec.js [hunk 12]
```diff
@@ -1,7 +1,13 @@
...
```
Comment (dor-itzhaki) [inline test/it/negative.spec.js]: why are we ignoring it ?

File: src_types/common/Buffer.d.ts [hunk 13]
```diff
@@ -0,0 +1,2 @@
+declare class Buffer extends Uint8Array{
```
Comment (dor-itzhaki) [inline src_types/common/Buffer.d.ts]: it's not really Buffer, maybe we should rename the file

File: test/scripts/selector-declaration-builder.test.js [hunk 14]
```diff
@@ -0,0 +1,100 @@
+const $wDeclarationBuilder = require("../../scripts/selector-declaration-builder");
+
+describe("define $w.d.ts", () => {
```
Comment (dor-itzhaki) [inline test/scripts/selector-declaration-builder.test.js]: do we really need these tests ?
we usually try to write tests that verify the actual product, in our case, making sure the user gets correct autocomplete.
i'd prefer not to maintain unit tests that check the internals.

File: package.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (dor-itzhaki) [reply package.json]: i understand the concern, but i think we shouldn't write code that does things we can do without writing any code. it's more work now, more maintenance and more room for bugs. so i don't think we should "generate" static files.

some other options:
1. put the generated files in their own subfolder
2. keep empty versions of the generated files in git with some placeholder like `// generated during build`
3. do the last thing @tomenden suggested - copying the static files to a dist folder where the generated files will be as well

File: scripts/selector-declaration-generator.js (hunk 5 above)
Comment (dor-itzhaki) [reply scripts/selector-declaration-generator.js]: why not use some unique temporary path for the cloned docs ?

**Comment by dor-itzhaki** [comment]
@agankarin i reviewed it partially, didn't really go into the generation logic, but i did add a few comments i consider to be important.

File: test/it/negative.spec.js (hunk 12 above)
Comment (agankarin) [reply test/it/negative.spec.js]: Adding a new class to the SDK(in this case AddressInput) breaks this test(generating the model, even in the old flow, breaks it). I added a new task to our current

File: src_types/common/Buffer.d.ts (hunk 13 above)
Comment (agankarin) [reply src_types/common/Buffer.d.ts]: I also, change it to an empty class(not extending any thing)

File: src_types/common/utilityTypes.d.ts (hunk 11 above)
Comment (agankarin) [reply src_types/common/utilityTypes.d.ts]: Well, actually I already did it in `selector-declaration-builder.js` but I just forgot to remove this file :)

File: .gitignore (hunk 9 above)
Comment (agankarin) [reply .gitignore]: temporary folder
I use `tmp` module instead

File: test/scripts/selector-declaration-builder.test.js (hunk 8 above)
Comment (agankarin) [reply test/scripts/selector-declaration-builder.test.js]: Oops, I removed this flow

File: package.json (hunk 4 above)
Comment (agankarin) [reply package.json]: So we go with the alternative and nicer solution, and created one generated file with all constants and other stuff :)

File: types/common/utilityTypes.d.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (agankarin) [reply types/common/utilityTypes.d.ts]: done

File: package.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (agankarin) [reply package.json]: so.. we go with copying the files from kind of src to types folder. The all `types` folder is ignored in `.gitignore`(I think is less confusing)
In the future, the all `types` would be generated(when we'll split the backend and frontend declaration, and the `index.js` would have different  `<reference path="` declaration)
Hope everyone like it :)

File: configs/tsconfig.backend.json (hunk 10 above)
Comment (agankarin) [reply configs/tsconfig.backend.json]: I added it back for now(as you said, it's not part of this task) but I spoke with the runtime time again and they said we don't want to let our user know the code is running in Node.js on propose and which Node.js modules the user can use. This is way we couldn't find this information in our articles.
Not sure if we need to deal with this question now, or we should just need to remove this support for now and add a task to our backlog(currently the online IDE doesn't support this syntax(`require`) and doesn't dive autocomplete for node.js module)

File: test/scripts/selector-declaration-builder.test.js (hunk 14 above)
Comment (agankarin) [reply test/scripts/selector-declaration-builder.test.js]: Let's talk about this(and other question like work flows, the QA goal in our team etc.) in our offsite next week. WDYT?

**Comment by agankarin** [comment]
@tomenden @dor-itzhaki I think I'd fix everything. Can you please have a final look?

File: test/scripts/services/$w.service.json [hunk 15]
```diff
@@ -0,0 +1,683 @@
+{ "name": "$w",
```
Comment (dor-itzhaki) [inline test/scripts/services/$w.service.json]: rename folder to fixtures ?

File: scripts/selector-declaration-builder.js [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+  const queryableType = getQueryableObjectType(services);
+  const wixElementSelectorType = getWixElementSelectorType();
+
+  return { queryableType, wixElementSelectorType, $wFunc, $wNamespace };
+}
+
+function $wDeclarationBuilder(services) {
+  const $wDeclaration = getAllDeclarations(services);
+
+  const intersectionArrayAndBaseDeclaration = `declare type ${INTERSECTION_ARRAY_AND_BASE_TYPE}<T, P> = {
```
Comment (dor-itzhaki) [inline scripts/selector-declaration-builder.js]: consider removing this type and adding the array option to the type map itself
